hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/RenameAtomicity.java

```diff
@@ -0,0 +1,380 @@
+    abfsClient.flush(blockList.getBytes(StandardCharsets.UTF_8),
+        path.toUri().getPath(), true, null, null, eTag, tracingContext);
+  }
+
+  @VisibleForTesting
+  public int preRename() throws IOException {
+    String makeRenamePendingFileContents = makeRenamePendingFileContents(
+        srcEtag);
+
+    try {
+      createRenamePendingJson(renameJsonPath,
+          makeRenamePendingFileContents.getBytes(StandardCharsets.UTF_8));
+      return makeRenamePendingFileContents.length();
+    } catch (IOException e) {
+      /*
+       * Scenario: file has been deleted by parallel thread before the RenameJSON
+       * could be written and flushed. In such case, there has to be one retry of
+       * preRename.
+       * ref: https://issues.apache.org/jira/browse/HADOOP-12678
+       * On DFS endpoint, flush API is called. If file is not there, server returns
+       * 404.
+       * On blob endpoint, flush API is not there. PutBlockList is called with
+       * if-match header. If file is not there, the conditional header will fail,
+       * the server will return 412.
+       */
+      if (isPreRenameRetriableException(e)) {
+        preRenameRetryCount++;
+        if (preRenameRetryCount == 1) {
+          return preRename();
+        }
+      }
+      throw e;
+    }
+  }
+
+  private boolean isPreRenameRetriableException(IOException e) {
+    AbfsRestOperationException ex;
+    while (e != null) {
+      if (e instanceof AbfsRestOperationException) {
+        ex = (AbfsRestOperationException) e;
+        return ex.getStatusCode() == HTTP_NOT_FOUND
+            || ex.getStatusCode() == HTTP_PRECON_FAILED;
+      }
+      e = (IOException) e.getCause();
+    }
+    return false;
+  }
+
+  public void postRename() throws IOException {
+    deleteRenamePendingJson();
+  }
+
+  private void deleteRenamePendingJson() throws AzureBlobFileSystemException {
+    try {
+      abfsClient.deleteBlobPath(renameJsonPath, null,
+          tracingContext);
+    } catch (AzureBlobFileSystemException e) {
+      if (e instanceof AbfsRestOperationException
+          && ((AbfsRestOperationException) e).getStatusCode()
+          == HTTP_NOT_FOUND) {
+        return;
+      }
+      throw e;
+    }
+  }
+
+
+  /**
+   * Return the contents of the JSON file to represent the operations
+   * to be performed for a folder rename.
+   *
+   * @return JSON string which represents the operation.
+   */
+  private String makeRenamePendingFileContents(String eTag) {
+    SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
+    sdf.setTimeZone(TimeZone.getTimeZone("UTC"));
+    String time = sdf.format(new Date());
+    if (StringUtils.isNotEmpty(eTag) && !eTag.startsWith("\"")
+        && !eTag.endsWith("\"")) {
+      eTag = quote(eTag);
+    }
+
+    // Make file contents as a string. Again, quote file names, escaping
+    // characters as appropriate.
+    String contents = "{\n"
+        + "  FormatVersion: \"1.0\",\n"
```

Comment:
FormatVersion, OldFolderName etc should all come from Constants

Reply:
Sure, will create  a new class for serialization and deserialization to json string. That would standardize the json.